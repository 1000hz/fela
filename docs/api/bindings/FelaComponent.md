# FelaComponent

FelaComponent is an alternative component to the [createComponent](createComponent.md)-HoC leveraging the render-props pattern. It uses [FelaTheme](FelaTheme.md) internally in order to access the theme directly.

## Props

| Property | Type | Default | Description |
| --- | --- | --- | --- |
| customClass | string | class(es) to prepend before the generated classes |
| style | [*StyleObject*](../../basics/Rules.md#styleobject)<br>*Function*| | Either a valid style object or a function of `theme` |
| rule | *Function*| | A function of `theme` and props|
| render | *string?*<br>*Function* | `div` | Either a render function or a string primitive to render into.<br>If passing a render function is receives the specified render interface. |

#### Render Interface
| Property | Type | Default | Description |
| --- | --- | --- | --- |
| className | *string* | | The class names for the rendered *style* object |
| theme | *Object* | `{}` | The theme object which is passed down via context |

## Imports
```javascript
import { FelaComponent } from 'react-fela'
import { FelaComponent } from 'preact-fela'
import { FelaComponent } from 'inferno-fela'
```

## Example
```javascript
<FelaComponent
  style={{
    backgroundColor: 'blue',
    color: 'red'
  }}
  render={({ className, theme }) => (
    <div className={className}>I am red on blue.</div>
  )}
/>
```

#### Generic Components
In order to create dynamically styled components using props, just like [createComponent](createComponent.md), we can create a component that itself renders a FelaComponent.

```javascript
const Button = ({ color, big = false, text }) => (
  <FelaComponent
    style={{
      backgroundColor: color,
      fontSize: big ? 18 : 15
    }}
    render={({ className }) => (
      <button className={className}>{text}</button>
    )}
  />
)
```

#### Using Theme
To access theme properties, we can simply pass a function of theme.

```javascript
<FelaComponent
  style={theme => ({
    backgroundColor: theme.bgPrimary,
    color: 'red'
  })}
  render={({ className, theme }) => (
    <div className={className}>I am red on {theme.bgPrimary}.</div>
  )}
/>
```

#### Style Rule as a Function of Props and Theme
Sometimes it is desirable to style a component as a function of both theme and
props. The `rule` prop takes a callback, and passes it an object with `theme`
and all props passed to FelaComponent except "style", "render" and "rule".

This provides an API that is both compatible with createComponent, and allows
using an externally defined function in such use cases. Hopefully, this can
be help performance in hot paths by not requiring a function to be created on
every render.


```javascript
const ruleFunction = ({ theme, bgc }) => ({
  backgroundColor: bgc || 'red',
  color: theme.bgPrimary,
})

<FelaComponent
  bgc='blue',
  rule={ruleFunction}
  render={({ className, theme }) => (
    <div className={className}>I am {theme.bgPrimary} on {bgc || 'red'}.</div>
  )}
/>
```

#### Shorthand Primitives
Instead of a render function, we can also specify a primitive element to render into.<br>
Children will automatically be passed down. If not specified at all, it will render into a `div`.

```javascript
<FelaComponent
  style={{
    backgroundColor: 'blue',
    color: 'red'
  }}
  render='span'
>
  I am red on blue
</FelaComponent>
```

#### Add Custom Classes
The common use case of needing to add a custom class to the generated ones, e.g.
for integration with 3rd party libraries, can be handled using the `customClass`
prop.

```javascript
<FelaComponent
  customClass="my-custom-class"
  style={{ color: 'red' }}
>
  I am red and have a custom class
</FelaComponent>
// <div class="my-custom-class a">
//  I am red and have a custom class
// </div>
```