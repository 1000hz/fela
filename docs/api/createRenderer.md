# `createRenderer(mountNode, [config])`

Creates a Fela renderer which renders your selectors, keyframes, fonts and static styles to the DOM. <br>It caches all rendered styles to be able to reuse them on future rendering cycles.<br>

## Arguments
1. `mountNode` (*[Element](https://developer.mozilla.org/en-US/docs/Web/API/Element)*): DOM node to render your CSS into. It should be a [`<style>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/style) element. Though in development, all other [HTMLElement](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement)s will work too, e.g. if you want to render the CSS into a [`<div>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/div) for debugging.
2. `config`(*Object?*): Optional renderer configuration. The most common use case adding [plugins](../advanced/Plugins.md) to process styles before they get rendered into the DOM. *(See [renderer configuration](Renderer.md#configuration) for further information)*.

### Returns
([Renderer](Renderer.md)): A Renderer instance bound to the `mountNode`.

### Example

```javascript
import { createRenderer } from 'fela'

// You would most likely add an existing <style>-element
// to your index.html and reference it with a special id
const mountNode = document.getElementById('stylesheet')

const rule = props => ({
  backgroundColor: 'red',
  fontSize: props.size,
  color: 'blue'
})

const renderer = createRenderer(mountNode)

renderer.render(rule, { size: '12px' }) // => c0 c0-dzm1d6


console.log(mountNode.textContent)
// .c0{background-color:red;color:blue}.c0-dzm1d6{font-size:12px}
```