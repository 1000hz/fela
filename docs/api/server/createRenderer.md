# `createRenderer([config])`

Creates a Fela renderer which silently renders your selectors, keyframes, fonts and static styles. <br>It caches all rendered styles to be able to reuse them on future rendering cycles.<br>

> Technically behaves exactly as the basic [`createRenderer(mountNode, [config])`](../createRenderer.md) does, except actually rendering the styles to a DOM node.

## Arguments
1. `config`(*Object?*): Optional renderer configuration. The most common use case adding [plugins](../advanced/Plugins.md) to process styles before they get rendered into the DOM. *(See [renderer configuration](Renderer.md#configuration) for further information)*.

### Returns
([Renderer](Renderer.md)): A silent Renderer instance.

### Example

> **Note**: Now we do not import from `fela` directly, but use `fela/server` instead, to address the server-side rendering package.

```javascript
import { createRenderer } from 'fela/server'

const rule = props => ({
  backgroundColor: 'red',
  fontSize: props.size,
  color: 'blue'
})

const renderer = createRenderer()

renderer.render(rule, { size: '12px' }) // => c0 c0-dzm1d6


console.log(renderer.renderToString())
// .c0{background-color:red;color:blue}.c0-dzm1d6{font-size:12px}
```
