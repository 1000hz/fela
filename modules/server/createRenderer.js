import Renderer from '../renderer/Renderer'

export default function createRenderer(config) {
  return new Renderer(config)
}