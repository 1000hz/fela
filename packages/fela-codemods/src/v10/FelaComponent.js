import getImportName from '../utils/getImportName'

const pkg = 'react-fela'
const component = 'FelaComponent'

export default function transformer(file, api, options) {
  const j = api.jscodeshift

  const importName = getImportName(j, file.source, pkg, component)

  // only transform if FelaComponent is imported from react-fela
  if (importName) {
    return j(file.source)
      .find(j.JSXElement)
      .forEach(path => {
        if (path.node.openingElement.name.name === importName) {
          const styleProp = path.node.openingElement.attributes.find(
            prop => prop.name.name === 'style'
          )

          const ruleProp = path.node.openingElement.attributes.find(
            prop => prop.name.name === 'rule'
          )

          // replace inline style expression from theme
          if (styleProp) {
            j(styleProp)
              .find(j.ArrowFunctionExpression)
              .forEach(path => {
                const param = path.node.params[0]

                if (param.type === 'Identifier') {
                  j(path).replaceWith(
                    j.arrowFunctionExpression(
                      [
                        j.objectPattern([
                          j.objectProperty(param, param, false, true),
                        ]),
                      ],
                      path.node.body
                    )
                  )

                  // make property truly shorthand
                  path.node.params[0].properties[0].shorthand = true
                }
              })
          }

          // rename rule to style
          if (ruleProp) {
            ruleProp.name.name = 'style'
          }
        }
      })
      .toSource()
  }

  return file.source
}
