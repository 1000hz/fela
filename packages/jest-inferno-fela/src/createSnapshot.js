import { render } from 'inferno'
import { createElement } from 'inferno-create-element'
import { Provider } from 'inferno-fela'

import { createSnapshotFactory } from 'jest-fela-bindings'

export default createSnapshotFactory(createElement, render, Provider)
