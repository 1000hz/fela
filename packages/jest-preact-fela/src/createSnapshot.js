import { createElement, render } from 'preact'
import { createRenderer } from 'fela'
import { RendererProvider, ThemeProvider } from 'preact-fela'

import { createSnapshotFactory } from 'jest-fela-bindings'

function renderToMarkup(component) {
  const div = document.createElement('div')

  render(component, div)

  return div.innerHTML
}

export default createSnapshotFactory(
  createElement,
  renderToMarkup,
  createRenderer(),
  RendererProvider,
  ThemeProvider
)
