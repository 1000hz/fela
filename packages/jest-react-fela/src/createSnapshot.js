import React from 'react'
import { render } from 'react-dom'
import { format } from 'prettier'
import HTMLtoJSX from 'htmltojsx'
import { renderToString } from 'fela-tools'
import { Provider as RendererProvider } from 'react-fela'

function newLine(str) {
  return '\n' + str
}

function formatCSS(css) {
  return format(css, { parser: 'css', useTabs: false, tabWidth: 2 })
}

function formatHTML(html) {
  const converter = new HTMLtoJSX({
    createClass: false,
  })

  const jsx = converter.convert(html)
  return format(jsx).replace(/[\\"]/g, '')
}

export default function createSnapshot(
  component,
  renderer,
  Provider = RendererProvider
) {
  const div = document.createElement('div')

  // reset renderer to have a clean setup
  renderer.clear()

  render(<Provider renderer={renderer}>{component}</Provider>, div)

  return (
    formatCSS(renderToString(renderer)) + '\n\n' + formatHTML(div.innerHTML)
  )
}
