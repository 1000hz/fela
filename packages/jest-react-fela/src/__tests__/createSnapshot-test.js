import React from 'react'

import createSnapshot from '../createSnapshot'

import FelaComponent from '../../../react-fela/src/FelaComponent'
import createRenderer from '../../../fela/src/createRenderer'
import webPreset from '../../../fela-preset-web/src/index'

describe('Creating Snapshots with Fela', () => {
  it('should return formatted html and css', () => {
    const renderer = createRenderer({
      plugins: webPreset,
    })

    const style = {
      display: 'flex',
      flex: '1 0 auto',
      ':hover': {
        color: 'blue',
      },
      '@media (min-width: 300px)': {
        backgroundColor: 'black',
        ':hover': {
          color: 'green',
        },
      },
    }

    expect(
      createSnapshot(<FelaComponent style={style} />, renderer)
    ).toMatchSnapshot()
  })

  it('should always use a clean setup', () => {
    const renderer = createRenderer({
      plugins: webPreset,
    })

    expect(
      createSnapshot(
        <FelaComponent style={{ color: 'red', backgroundColor: 'blue' }} />,
        renderer
      )
    ).toMatchSnapshot()
    expect(
      createSnapshot(
        <FelaComponent style={{ color: 'red', backgroundColor: 'green' }} />,
        renderer
      )
    ).toMatchSnapshot()
  })
})
