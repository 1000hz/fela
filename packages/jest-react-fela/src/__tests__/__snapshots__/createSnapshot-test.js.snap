// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Creating Snapshots with Fela should always use a clean setup 1`] = `
".a {
  color: red;
}
.b {
  background-color: blue;
}


<div className=a b />;
"
`;

exports[`Creating Snapshots with Fela should always use a clean setup 2`] = `
".a {
  color: red;
}
.b {
  background-color: green;
}


<div className=a b />;
"
`;

exports[`Creating Snapshots with Fela should return both html and css 1`] = `
".a {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
}
.b {
  -webkit-flex: 1 0 auto;
  flex: 1 0 auto;
}
.e:hover {
  color: blue;
}
@media (min-width: 300px) {
  .c {
    background-color: black;
  }
  .d:hover {
    color: green;
  }
}


<div className=a b c d e />;
"
`;

exports[`Creating Snapshots with Fela should return formatted html and css 1`] = `
".a {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
}
.b {
  -webkit-flex: 1 0 auto;
  flex: 1 0 auto;
}
.e:hover {
  color: blue;
}
@media (min-width: 300px) {
  .c {
    background-color: black;
  }
  .d:hover {
    color: green;
  }
}


<div className=a b c d e />;
"
`;
